relay-server/src/metrics_extraction/sessions/types.rs

```diff
@@ -1,15 +1,156 @@
 use std::collections::BTreeMap;
+use std::fmt::{self, Display};
 
-use relay_common::{UnixTimestamp, Uuid};
+use relay_common::{MetricUnit, UnixTimestamp, Uuid};
 use relay_general::protocol::{
     AbnormalMechanism, SessionAttributes, SessionErrored, SessionLike, SessionStatus,
 };
-use relay_metrics::{Metric, MetricNamespace, MetricUnit, MetricValue};
+use relay_metrics::{
+    CounterType, Metric, MetricNamespace, MetricResourceIdentifier, MetricType, MetricValue,
+};
+
+use crate::metrics_extraction::IntoMetric;
+
+/// Enumerates the metrics extracted from session payloads.
+#[derive(Clone, Debug, PartialEq)]
+pub enum SessionMetric {
+    /// The number of sessions collected in the given time frame.
+    Session {
+        counter: CounterType,
+        tags: SessionSessionTags,
+    },
+    /// The number of unique session users for the given time frame.
+    User {
+        distinct_id: String,
+        tags: SessionUserTags,
+    },
+    /// The number of sessions that errored in the given time frame.
+    ///
+    /// Because multiple session updates can be received for the same session ID,
+    /// this is collected as a [`MetricType::Set`] metric rather than a simple counter.
+    Error {
+        session_id: Uuid,
+        tags: SessionErrorTags,
+    },
+}
+
+impl SessionMetric {
+    fn ty(&self) -> MetricType {
+        match self {
+            SessionMetric::Session { .. } => MetricType::Counter,
+            SessionMetric::Error { .. } => MetricType::Set,
+            SessionMetric::User { .. } => MetricType::Set,
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct SessionErrorTags {
+    common_tags: CommonTags,
+}
```

Comment:
In this case I would take a shortcut and declare `tags: CommonTags` in the `Session` variant.